pkg/engine/internal/planner/logical/planner.go

```diff
@@ -186,101 +188,179 @@ func buildPlanForLogQuery(
+	}
 
-			switch e.Operation {
-			case syntax.OpRangeTypeCount:
-				rangeAggType = types.RangeAggregationTypeCount
-			case syntax.OpRangeTypeSum:
-				rangeAggType = types.RangeAggregationTypeSum
-			//case syntax.OpRangeTypeMax:
-			//	rangeAggType = types.RangeAggregationTypeMax
-			//case syntax.OpRangeTypeMin:
-			//	rangeAggType = types.RangeAggregationTypeMin
-			default:
-				err = errUnimplemented
-				return false
-			}
+	var rangeAggType types.RangeAggregationType
+	switch e.Operation {
+	case syntax.OpRangeTypeCount:
+		rangeAggType = types.RangeAggregationTypeCount
+	case syntax.OpRangeTypeSum:
+		rangeAggType = types.RangeAggregationTypeSum
+	//case syntax.OpRangeTypeMax:
+	//	rangeAggType = types.RangeAggregationTypeMax
+	//case syntax.OpRangeTypeMin:
+	//	rangeAggType = types.RangeAggregationTypeMin
+	default:
+		return nil, errUnimplemented
+	}
 
-			rangeInterval = e.Left.Interval
-			return false // do not traverse log range query
+	rangeAggregation := &RangeAggregation{
+		Table: logQuery,
 
-		case *syntax.VectorAggregationExpr:
-			// `without()` grouping is not supported.
-			if e.Grouping != nil && e.Grouping.Without {
-				err = errUnimplemented
-				return false
-			}
+		Operation:     rangeAggType,
+		PartitionBy:   nil,
+		Start:         params.Start(),
+		End:           params.End(),
+		Step:          params.Step(),
+		RangeInterval: rangeInterval,
+	}
 
-			switch e.Operation {
-			//case syntax.OpTypeCount:
-			//	vecAggType = types.VectorAggregationTypeCount
-			case syntax.OpTypeSum:
-				vecAggType = types.VectorAggregationTypeSum
-			//case syntax.OpTypeMax:
-			//	vecAggType = types.VectorAggregationTypeMax
-			//case syntax.OpTypeMin:
-			//	vecAggType = types.VectorAggregationTypeMin
-			default:
-				err = errUnimplemented
-				return false
-			}
+	return rangeAggregation, nil
+}
 
-			groupBy = make([]ColumnRef, 0, len(e.Grouping.Groups))
-			for _, group := range e.Grouping.Groups {
-				groupBy = append(groupBy, *NewColumnRef(group, types.ColumnTypeAmbiguous))
-			}
+func walkVectorAggregation(e *syntax.VectorAggregationExpr, params logql.Params) (Value, error) {
+	// `without()` grouping is not supported.
+	if e.Grouping != nil && e.Grouping.Without {
+		return nil, errUnimplemented
+	}
 
-			return true
-		default:
-			err = errUnimplemented
-			return false // do not traverse children
-		}
-	})
+	left, err := walk(e.Left, params)
 	if err != nil {
 		return nil, err
 	}
 
-	if rangeAggType == types.RangeAggregationTypeInvalid || vecAggType == types.VectorAggregationTypeInvalid {
+	var vecAggType types.VectorAggregationType
+	switch e.Operation {
+	//case syntax.OpTypeCount:
+	//	vecAggType = types.VectorAggregationTypeCount
+	case syntax.OpTypeSum:
+		vecAggType = types.VectorAggregationTypeSum
+	//case syntax.OpTypeMax:
+	//	vecAggType = types.VectorAggregationTypeMax
+	//case syntax.OpTypeMin:
+	//	vecAggType = types.VectorAggregationTypeMin
+	default:
 		return nil, errUnimplemented
 	}
```

Comment:
I know that you did not change anything here, but could we follow the pattern that you introduced and move this into a separate function `convertVectorAggType(op string) types.VectorAggregationType`?

Same for the RangeAggregationType